gaps/gap-25_golem_certificates/gap-25_golem_certificates.md

```diff
@@ -0,0 +1,181 @@
+
+## Rationale
+
+### Why can't we wait with introducing permissions?
+
+Any future releases introducing sensitive features will put Providers in danger, because the same certificates could be used to sign new features.
+
+### Are there any existing x509 extensions, that we could use?
+
+Although there are extensions serving similar purpose of defining, what certificate can be used for, we have no control over these extensions and can't define our own list of values.
+
+### Why we need `all` allowance?
+
+Since list of allowances is not exhaustive, we don't want to be forced to change root certificate, when adding new features.
+We must keep in mind that `all` allowance should be used sparingly only for top level certificates. For sure we shouldn't sign any certificate with `all` allowance for external parties.
+
+## Backwards Compatibility
+
+Since outbound network isn't release yet, we don't have any problems with backward compatibility.
+This GAP aims to solve backward compatibility and security problems, that could arise in the future.
+
+## Test Cases
+
+## [Optional] Reference Implementation
+
+### Generating openssl certificate
+
+You can check [here](file://cert.confi) example config, which can be passed to openssl to generate certificate containing expected fields. To generate certificate run:
+
+`openssl req -x509 -sha256 -days 1825 -newkey rsa:4096 -keyout rootCA.key -out rootCA.crt -config cert.config  --extensions golem_network_certificates`
+
+than check if certificate contains expected fields with:
+
+`openssl x509 -text -noout -in rootCA.crt`
+
+Output looks like this:
+
+```
+...
+        X509v3 extensions:
+            1.3.6.1.4.1.60000.1.1: 
+                0...all..manifest-outbound
+...
+```
+
+### Provider Agent changes
+
+Provider agent supports using certificates without `allowance` extensions.
+We need following changes:
+
+- Certificates chain permissions validation
+- Check outbound allowance during negotiations (currently signature is enough)
+
+No changes on Requestor side are expected.
+
+## Security Considerations
+
+### Old yagna versions
+
+If we don't implement certificates `allowance` before releasing outbound network, Providers could accept Manifest signatures signed by certificates,
+that don't have `manifest-outbound` allowance. This doesn't hurt, as long as we don't introduce new sensitive features.
+
+### Private certificates keys management
+
+We need to consider security issues related to storing and using Root Golem Certificate and important intermediate certificates.
+We need internal company policies, who has permissions to decide about signing using certain certificates and who has access to them.
+
+### Root Golem certificate
+
+Root certificate is always self-signed certificate. We could sign it using [Golem Multisig](https://etherscan.io/address/0x7da82c7ab4771ff031b66538d2fb9b0b047f6cf9) contract to allow anyone validate it's authenticity.
```

Comment:
I strongly recommend leveraging a well-established public CA as issuer of Golem Root (or in this case - Golem Intermediate) certificates.  
We do not want to be a public CA in full sense, and we don't want to be legally liable for faults in managing the Root CA certs. Buying a Root/Intermediate cert from a public CA delegates risk away from GolemFactory, and will make Golem certificates be more accepted by the users in the "industry" (esp. the enterprise users). 
A concise article to back the above comment can be found here: https://www.thesslstore.com/blog/how-to-become-a-certificate-authority/

Reply:
I agree it is better to delegate reponsibility to some Public CA but it creates a question whether we should not delegate responsibility of keeping Public CA root certificates up to date to user/OS and to just take them from OS? It will also make easier for an app author to distribute their intermediate certificate signed by Public CA because there will be no necessity to import (and prepare) a whole chain.
If we will stick to internal certificate storage using Public CA will make having a proper certificate revocation handling in Provider's keystore an absolute must (there is already a task for it).

Reply:
On second thought I see a problem.
In some way we are turning x509 certificates into an authorization method.
The way we wanted to allow providers to control whom to trust or not was by importing App author's public certificates into provider's keystore. Right now we do not verify whether imported certificate (or chain) meets any criteria, we just check if it is valid and it gets saved inside a keystore. The moment Public CA certs will end up in Provider's keystore any Demand with manifest signed directly by Public CA certificate will get accepted. Anyone can aquire certificate signed by same Public CA so suddenly Provider starts to trust everyone.
Verifying `GolemCertificatesAllowance` in a certificate incoming with Demand will not help. What would need to be verified will be `GolemCertificatesAllowance` of the certificate signing the certificate incoming in a Demand (lets say `app-author` allowance). Alternatively we could just check if the fields like company name are the same between the certificate incoming in Demand and the certificate signing incoming one.

Reply:
@stranger80 After reading your link, I think, I imagined our role rather as private CA not public CA.
I wanted self-signed certificate, because I wasn't not sure if any CA authority will sign our certificate, if we are adding our custom extensions of unknown (from their perspective) purpose.

On the other side: does having self-signed certificate make us immediately public CA?

And my other concern is: although we could be signed by public CA, we can't have our parent certificate on Provider, because it won't have any `GolemCertificatesAllowance`, so the whole chain would be rejected by Provider.
This relates to your comment either @pwalski

Reply:
The second thing: I talked with @prekucki and we can't sign our certificate using Multisig anyway, because signature comes from only one holder and contract logic ensures, that all holders agreed to sign.

Reply:
And the last thing: I'm wondering if we shouldn't use blockchain for storing/handling these certificates.